src/libtls/tls_hkdf.c

```diff
@@ -0,0 +1,520 @@
+	 */
+	hash_algorithm_t hash_algorithm;
+
+	/**
+	 * Length of this hash output in bytes.
+	 */
+	size_t hash_length;
+
+	/**
+	 * Preshared key IKM
+	 */
+	chunk_t *psk;
+
+	/**
+	 * (EC)DHE IKM
+	 */
+	chunk_t *shared_secret;
+
+	/**
+	 * Pseudorandom function used.
+	 */
+	prf_t *prf;
+
+	/**
+	 * Pseudorandom plus function used.
+	 */
+	prf_plus_t *prf_plus;
+
+	/**
+	 * Hasher used.
+	 */
+	hasher_t *hasher;
+
+	/**
+	 * Salt used.
+	 */
+	chunk_t salt;
+
+	/**
+	 * IKM used.
+	 */
+	chunk_t ikm;
+
+	/**
+	 * PRK used.
+	 */
+	chunk_t prk;
+
+	/**
+	 * Info used.
+	 */
+	chunk_t info;
+
+	/**
+	 * OKM used.
+	 */
+	chunk_t okm;
+
+	/**
+	 * Length of the desired key output.
+	 */
+	size_t L;
+};
+
+static char *hkdf_labels[] = {
+	"tls13 ext binder",
+	"tls13 res binder",
+	"tls13 c e traffic",
+	"tls13 e expt master",
+	"tls13 c hs traffic",
+	"tls13 s hs traffic",
+	"tls13 c ap traffic",
+	"tls13 s ap traffic",
+	"tls13 exp master",
+	"tls13 res master",
+};
+
+/* HKDF-Extract(this->salt, this->ikm) -> this->prk */
+static bool extract(private_tls_hkdf_t *this, chunk_t *salt, chunk_t *ikm, chunk_t *prk)
+{
+	if (!this->prf->set_key(this->prf, *salt))
+	{
+		DBG1(DBG_TLS, "unable to set PRF salt");
+		return FALSE;
+	}
+	chunk_clear(prk);
+	if(!this->prf->allocate_bytes(this->prf, *ikm, prk))
+	{
+		DBG1(DBG_TLS, "unable to allocate PRF space");
+		return FALSE;
+	}
+
+	DBG3(DBG_TLS, "PRK: %B", prk);
+
+	return TRUE;
+}
+
+/* HKDF-Expand(this->prk, this->info, this->L) -> this->okm */
+static bool expand(private_tls_hkdf_t *this, chunk_t *prk, chunk_t *info, size_t lenght, chunk_t *okm)
+{
+	if (!this->prf->set_key(this->prf, *prk))
+	{
+		DBG1(DBG_TLS, "unable to set PRF PRK");
+		chunk_free(&this->info);
+		return FALSE;
+	}
+	DESTROY_IF(this->prf_plus);
+	this->prf_plus = prf_plus_create(this->prf, TRUE, *info);
+	chunk_clear(&this->okm);
+	if (!this->prf_plus->allocate_bytes(this->prf_plus, lenght, okm))
+	{
+		DBG1(DBG_TLS, "unable to allocate PRF plus space");
+		chunk_clear(okm);
+		chunk_free(&this->info);
+		return FALSE;
+	}
+	chunk_free(&this->info);
+
+	DBG3(DBG_TLS, "OKM: %B", okm);
+
+	return TRUE;
+}
+
+static bool expand_label(private_tls_hkdf_t *this, chunk_t *secret, chunk_t *label, chunk_t *context, uint16_t length, chunk_t *key)
+{
+	if (label->len < 7 || label->len > 255 ||
+		context->len < 0 || context->len > 255)
+	{
+		chunk_free(context);
+		return FALSE;
+	}
+
+	/* HKDFLabel */
+	size_t offset = 0;
+	size_t chunk_len = 2 + 1 + label->len + 1 + context->len;
+	this->info = chunk_alloc(chunk_len);
+	this->info.ptr[1] = length & 0xFF;
+	this->info.ptr[0] = length >> 8;
+	this->info.ptr[2] = label->len;
+	offset += 3;
+	memcpy(&(this->info).ptr[offset], label->ptr, label->len);
+	offset += label->len;
+	this->info.ptr[offset] = context->len;
+	offset += 1;
+	memcpy(&(this->info).ptr[offset], context->ptr, context->len);
+	offset += context->len;
+
+	expand(this, secret, &this->info, length, key);
+
+	return TRUE;
```

Comment:
Check the return value and pass it to the caller. And with the writer change above this would result in something  like this:
```suggestion
	success = expand(this, secret, writer->get_buf(writer), length, key);
	writer->destroy(writer);
	return success;
```